// This proto file defines all services (AuthService and ChatService) and message types for all the request
// and response types used in our service methods.

// Command to generate client + server stub code:
// python -m grpc_tools.protoc -I protos --python_out=. --pyi_out=. --grpc_python_out=. protos/chat.proto

syntax = "proto3";
package chatservice;

// AuthService handles register, login, and delete account requests
service AuthService {
  rpc Login (LoginRequest) returns (LoginResponse) {}
  rpc Register (RegisterRequest) returns (RegisterResponse) {}
  rpc DeleteAccount (DeleteRequest) returns (DeleteResponse) {}
}

// ChatService handles send/receive message requests and getting all the current users
service ChatService {
  rpc SendMessage (SendRequest) returns (SendResponse) {}
  rpc ReceiveMessage (User) returns (AllChats) {}
  rpc GetUsers (Empty) returns (AllUsers) {}
}

message Empty {}

message User{
  string username = 1;
}

message LoginRequest {
  string username = 1;
  string password = 2;
  bool is_client = 3;
}

message LoginResponse {
  bool success = 1;
  string message = 2;
}

message RegisterRequest {
  string username = 1;
  string password = 2;
  bool is_client = 3;
  uint64 counter = 4;
}

message RegisterResponse {
  bool success = 1;
  string message = 2;
}

message DeleteRequest {
  string username = 1;
  string password = 2;
  bool is_client = 3;
  uint64 counter = 4;
}

message DeleteResponse {
  bool success = 1;
  string message = 2;
}

message SendRequest {
  string sender = 1;
  string receiver = 2;
  string body = 3;
  bool is_client = 4;
  uint64 counter = 5;
}

message SendResponse {
  bool success = 1;
  string message = 2;
}

message AllChats {
  repeated string chats = 1;
}

message AllUsers {
  repeated string users = 1;
}
